src/core/lib/security/credentials/fake/fake_credentials.cc

```diff
@@ -33,6 +33,11 @@
 #define GRPC_ARG_FAKE_SECURITY_EXPECTED_TARGETS \
   "grpc.fake_security.expected_targets"
 
+typedef struct {
+  grpc_server_credentials base;
+  int add_sniffing_handshaker;
```

Comment:
What's this sniffing handshaker stuff for?  It's not mentioned in the PR description, so it looks unrelated.

Reply:
Updated the PR description; I added this only to put a test in this repo that could notice this functionality missing. That said I'm not completely sure if it's worth having the new test here.

Reply:
Okay, I think I see what what the goal is here.  But in that case, I don't think this implementation is correct.

In order to test that the security handshaker is doing the right thing, we need to inject a new non-security handshaker into the handshaker list that will run before the security handshaker does, read some bytes into `h->args->read_buffer`, and then hand off control to the security handshaker.  That way, we can verify that the security handshaker will process the bytes read for it by the preceding handshaker.

As currently written, this code will actually replace the security handshaker with this new handshaker.  In other words, this is actually testing that the **transport** is using `h->args->read_buffer`, not that the security handshaker is -- in fact, the security handshaker is not actually running at all in this test.

So here's what I suggest:
- Move this new handshaker into a test.  It should not be defined outside of the test, because it won't be used anywhere else.
- The new handshaker should not be injected using security credentials, because that will replace the security handshaker.  Instead, we want to inject the new handshaker prior to the security handshaker.  You can do that using a handshaker factory; as an example, see https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/http_connect_handshaker.cc#L373.

Once you do that, I think you can use your new test handshaker with SSL credentials and verify that the TLS handshake still succeeds.

Reply:
Might be wrong but IIC this test is currently running the sniffing handshaker before but *with* the tsi handshaker; I changed the security connector's handshaker adder to do optionally add the sniffing handshaker in front of the tsi handshaker.

Thanks for suggestions though, changing this around to follow that, which sounds cleaner.